README.md
gtin
===

[![npm version](https://img.shields.io/npm/v/gtin.svg?style=flat-square)](https://npmjs.com/package/gtin)
[![javascript standard style](https://img.shields.io/badge/code%20style-standard-blue.svg?style=flat-square)](http://standardjs.com/)
[![travis build](https://img.shields.io/travis/xbpf/gtin/master.svg?style=flat-square)](https://travis-ci.org/xbpf/gtin)
[![coveralls coverage](https://img.shields.io/coveralls/xbpf/gtin.svg?style=flat-square)](https://coveralls.io/github/xbpf/gtin)
[![david dependencies](https://david-dm.org/xbpf/gtin.svg?style=flat-square)](https://david-dm.org/xbpf/gtin)
[![david dev dependencies](https://david-dm.org/xbpf/gtin/dev-status.svg?style=flat-square)](https://david-dm.org/xbpf/gtin)


GTIN (UPC, EAN, ITF, etc.) utilities.

`npm install gtin`

---

* [gtin.isGTIN(barcode)](#user-content-gtin-isGTIN)
* [gtin.isValid(barcode)](#user-content-gtin-isValid)
* [gtin.minify(barcode)](#user-content-gtin-minify)
* [gtin.getFormat(barcode)](#user-content-gtin-getFormat)
* [gtin.getRealFormat(barcode)](#user-content-gtin-getRealFormat)
* [gtin.upcE.compress(barcode)](#user-content-gtin-upcE-compress)
* [gtin.upcE.expand(barcode)](#user-content-gtin-upcE-expand)

---

<a id='gtin-isGTIN'></a>
`gtin.isGTIN(barcode)`
---

Returns true or false, depending on if the string given is a GTIN barcode.
Throws an error if an empty string or anything other than a string is provided.

**NOTE**: This does not validate the code by check digit. Validation is done
with `gtin.isValid`.

```js
import { isGTIN } from 'gtin'

isGTIN('1234')      // false
isGTIN('12341238')  // true
isGTIN('')          // Error thrown
isGTIN(123)         // Error thrown
```

<a id='gtin-isValid'></a>
`gtin.isValid(barcode)`
---

Validates a GTIN (14, 13, 12, or 8-digit) barcode by check digit. Barcode must
be a string.

To validate a UPC-E barcode, expand it first: `isValid(upcE.expand('01278906'))`

```js
import { isValid } from 'gtin'

isValid('12341238')       // true
isValid('1234123412344')  // true
isValid('12341234123413') // true
isValid('012000007897')   // true
isValid('012000007896')   // false
isValid('abc')            // Error thrown
isValid(123)              // Error thrown
isValid('123')            // Error thrown
```

<a id='gtin-minify'></a>
`gtin.minify(barcode)`
---

Minifies GTIN to smallest possible representation, by stripping as many leading
zeroes as possible. Does not compress to UPC-E.

```js
import { minify } from 'gtin'

minify('00000012341238')  // '12341238'
minify('0000012341238')   // '12341238'
minify('01234123412344')  // '1234123412344
minify('001234123412344') // Error thrown
minify('abc')             // Error thrown
minify(123)               // Error thrown
minify('123')             // Error thrown
```

<a id='gtin-getFormat'></a>
`gtin.getFormat(barcode)`
---

Gets the format of the given barcode. Does not validate checksum.

```js
import { getFormat } from 'gtin'

getFormat('12341238')       // 'GTIN-8'
getFormat('123412341234')   // 'GTIN-12'
getFormat('1234123412344')  // 'GTIN-13'
getFormat('01234123412344') // 'GTIN-14'
getFormat('123412381')      // Error thrown
getFormat('abc')            // Error thrown
getFormat(123)              // Error thrown
getFormat('123')            // Error thrown
```

<a id='gtin-getRealFormat'></a>
`gtin.getRealFormat(barcode)`
---

Gets the real format of the given barcode, by minifying it first.

```js
import { getRealFormat } from 'gtin'

getRealFormat('1234123412344')  // 'GTIN-13'
getRealFormat('01234123412344') // 'GTIN-13'
getRealFormat('123412381')      // Error thrown
getRealFormat('abc')            // Error thrown
getRealFormat(123)              // Error thrown
getRealFormat('123')            // Error thrown
```

<a id='gtin-upcE-compress'></a>
`gtin.upcE.compress(barcode)`
---

Compress a UPC-A barcode to an 8-digit UPC-E barcode. Does not validate
code by check digit. Barcode must be a string.

* 12-digit UPC-A: Number system and check digits are taken into account.
* 11-digit UPC-A: Number system 0 is assumed. Check digit is taken into account.
* 10-digit UPC-A: Number system 0 is assumed. Check digit is generated.

```js
import { upcE } from 'gtin'

upcE.compress('1200000789')   // '01278907'
upcE.compress('12000007897')  // '01278907'
upcE.compress('012000007897') // '01278907'
upcE.compress('012000007896') // '01278906'
upcE.compress('012345678905') // null
upcE.compress(123)            // Error thrown
upcE.compress('123')          // Error thrown
upcE.compress('abc')          // Error thrown
```

<a id='gtin-upcE-expand'></a>
`gtin.upcE.expand(barcode)`
---

Expands a UPC-E barcode to a 12-digit UPC-A barcode. Does not validate
code by check digit. Barcode must be a string.

* 8-digit UPC-E: Number system and check digits are taken into account.
* 7-digit UPC-E: Number system 0 is assumed. Check digit is taken into account.
* 6-digit UPC-E: Number system 0 is assumed. Check digit is generated.

```js
import { upcE } from 'gtin'

upcE.expand('127890')    // '012000007897'
upcE.expand('1278907')   // '012000007897'
upcE.expand('01278907')  // '012000007897'
upcE.expand('01278906')  // '012000007896'
upcE.expand('123412341') // Error thrown
upcE.expand(123)         // Error thrown
upcE.expand('123')       // Error thrown
upcE.expand('abc')       // Error thrown
```


CHANGELOG.md
`master`
===

No changes yet.

--------------------------------------------------------------------------------

`1.0.2`
===

## Other changes

- Fixed README to have correct examples (again).

--------------------------------------------------------------------------------

`1.0.1`
===

## Other changes

- Fixed README to have correct examples.

--------------------------------------------------------------------------------

`1.0.0`
===

## Breaking changes

- Minimum supported Node.js version is now `10`.

  Note that this library may still work with older versions, but no
  effort will be made to fix this library to support those versions.

- Support for browsers that don't support `String.padStart` (e.g. all IE and opera mini versions) has been removed.

  Use [Babel](https://babeljs.io/) if you need to support older browsers.

- API rename
  - `gtin.validate` is now `gtin.isValid`
  - `gtin.upce` is now `gtin.upcE`

    Alternatively, you can also import the UPC-E functions directly:
    `import upcE from 'gtin/lib/upc-e'`

## Other changes

- Typescript support

  `.d.ts` declaration files are now included in the published package.

--------------------------------------------------------------------------------

`0.3.0`
===

Initial release.


test/test-upce.js
'use strict'

import test from 'ava'

import { upce } from '../'

test('gtin.upce.compress(barcode) (compressable)', t => {
  t.plan(7)

  const cases = [
    ['1200000789', '01278907'],
    ['12000007897', '01278907'],
    ['012000007897', '01278907'],
    ['012000007896', '01278906'],
    ['012300000890', '01238930'],
    ['012910000090', '01291940'],
    ['012911000050', '01291150']
  ]

  for (const testCase of cases) {
    t.is(
      upce.compress(testCase[0]),
      testCase[1]
    )
  }
})

test('gtin.upce.compress(barcode) (uncompressable)', t => {
  t.plan(3)

  const cases = [
    '1234567890',
    '12345678905',
    '012345678905'
  ]

  for (const testCase of cases) {
    t.is(
      upce.compress(testCase),
      null
    )
  }
})

test('gtin.upce.compress(barcode) (invalid)', t => {
  t.plan(3)

  const cases = [
    [123, 'Barcode must be a string'],
    ['123', 'Barcode is not of a valid format'],
    ['abc', 'Barcode is not of a valid format']
  ]

  for (const testCase of cases) {
    try {
      upce.compress(testCase[0])
      t.fail('Error was not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})

test('gtin.upce.expand(barcode) (expandable)', t => {
  t.plan(7)

  const cases = [
    ['127890', '012000007897'],
    ['1278907', '012000007897'],
    ['01278907', '012000007897'],
    ['01278906', '012000007896'],
    ['01238930', '012300000890'],
    ['01291940', '012910000090'],
    ['01291150', '012911000050']
  ]

  for (const testCase of cases) {
    t.is(
      upce.expand(testCase[0]),
      testCase[1]
    )
  }
})

test('gtin.upce.expand(barcode) (invalid)', t => {
  t.plan(4)

  const cases = [
    [123, 'Barcode must be a string'],
    ['123412341', 'Barcode is not of a valid format'],
    ['123', 'Barcode is not of a valid format'],
    ['abc', 'Barcode is not of a valid format']
  ]

  for (const testCase of cases) {
    try {
      upce.expand(testCase[0])
      t.fail('Error was not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})


test/test-index.js
'use strict'

import test from 'ava'

import gtin from '../'

test('gtin.validate(barcode) (valid)', t => {
  t.plan(4)

  const cases = [
    '12341238',
    '012000007897',
    '1234123412344',
    '12341234123413'
  ]

  for (const testCase of cases) {
    t.is(gtin.validate(testCase), true)
  }
})

test('gtin.validate(barcode) (invalid)', t => {
  t.plan(4 * 9)

  const cases = [
    ['1234123', 8],
    ['01200000789', 7],
    ['123412341234', 4],
    ['1234123412341', 3]
  ]

  for (const testCase of cases) {
    for (let i = 0; i < 10; i++) {
      if (i === testCase[1]) continue
      t.is(gtin.validate(testCase[0] + String(i)), false)
    }
  }
})

test('gtin.validate(barcode) (error)', t => {
  t.plan(3)

  const cases = [
    [123, 'Barcode must be a string'],
    ['abc', 'Barcode is not of a valid format'],
    ['123', 'Barcode is not of a valid format']
  ]

  for (const testCase of cases) {
    try {
      gtin.validate(testCase[0])
      t.fail('Error not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})

test('gtin.minify(barcode) (minified)', t => {
  t.plan(5)

  const cases = [
    ['00000012341238', '12341238'],
    ['0000012341238', '12341238'],
    ['000012341238', '12341238'],
    ['12341238', '12341238'],
    ['01234123412344', '1234123412344']
  ]

  for (const testCase of cases) {
    t.is(
      gtin.minify(testCase[0]),
      testCase[1]
    )
  }
})

test('gtin.minify(barcode) (error)', t => {
  t.plan(4)

  const cases = [
    [123, 'Barcode must be a string'],
    ['abc', 'Barcode is not of a valid format'],
    ['123', 'Barcode is not of a valid format'],
    ['001234123412344', 'Barcode is not of a valid format']
  ]

  for (const testCase of cases) {
    try {
      gtin.minify(testCase[0])
      t.fail('Error not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})

test('gtin.getFormat(barcode) (ok)', t => {
  t.plan(4)

  const cases = [
    ['12341238', 'GTIN-8'],
    ['123412341234', 'GTIN-12'],
    ['1234123412344', 'GTIN-13'],
    ['01234123412344', 'GTIN-14']
  ]

  for (const testCase of cases) {
    t.is(
      gtin.getFormat(testCase[0]),
      testCase[1]
    )
  }
})

test('gtin.getFormat(barcode) (error)', t => {
  t.plan(4)

  const cases = [
    [123, 'Barcode must be a string'],
    ['abc', 'Barcode is not of a valid format'],
    ['123', 'Barcode is not of a valid format'],
    ['123412381', 'Barcode is not of a valid format']
  ]

  for (const testCase of cases) {
    try {
      gtin.getFormat(testCase[0])
      t.fail('Error not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})

test('gtin.getRealFormat(barcode) (ok)', t => {
  t.plan(2)

  const cases = [
    ['1234123412344', 'GTIN-13'],
    ['01234123412344', 'GTIN-13']
  ]

  for (const testCase of cases) {
    t.is(
      gtin.getRealFormat(testCase[0]),
      testCase[1]
    )
  }
})

test('gtin.getRealFormat(barcode) (error)', t => {
  t.plan(4)

  const cases = [
    [123, 'Barcode must be a string'],
    ['abc', 'Barcode is not of a valid format'],
    ['123', 'Barcode is not of a valid format'],
    ['123412381', 'Barcode is not of a valid format']
  ]

  for (const testCase of cases) {
    try {
      gtin.getRealFormat(testCase[0])
      t.fail('Error not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})

test('gtin.isGTIN(barcode) (ok)', t => {
  t.plan(8)

  const cases = [
    ['abcdabcdabcd', false],
    ['1234', false],
    ['123412341', false],
    ['12341234123412341', false],
    ['12341234', true],
    ['123412341234', true],
    ['1234123412343', true],
    ['12341234123434', true]
  ]

  for (const testCase of cases) {
    t.is(
      gtin.isGTIN(testCase[0]),
      testCase[1]
    )
  }
})

test('gtin.isGTIN(barcode) (error)', t => {
  t.plan(4)

  const cases = [
    [123, 'Barcode must be a string'],
    [{}, 'Barcode must be a string'],
    [undefined, 'Barcode must be a string'],
    ['', 'Barcode must not be an empty string']
  ]

  for (const testCase of cases) {
    try {
      gtin.isGTIN(testCase[0])
      t.fail('Error not thrown')
    } catch (err) {
      t.is(err.message, testCase[1])
    }
  }
})


lib/__tests__/upce.test.js
const upce = require('../upce')

test.each([
  ['1200000789', '01278907'],
  ['12000007897', '01278907'],
  ['012000007897', '01278907'],
  ['012000007896', '01278906'],
  ['012300000890', '01238930'],
  ['012910000090', '01291940'],
  ['012911000050', '01291150']
])('gtin.upce.compress(%p) (compressable: %p)', (originalBarcode, compressedBarcode) => {
  expect(upce.compress(originalBarcode)).toBe(compressedBarcode)
})

test.each([
  ['1234567890'],
  ['12345678905'],
  ['012345678905']
])('gtin.upce.compress() (uncompressable)', (barcode) => {
  expect(upce.compress(barcode)).toBe(null)
})

test.each([
  [123, 'Barcode must be a string'],
  ['123', 'Barcode is not of a valid format'],
  ['abc', 'Barcode is not of a valid format']
])('gtin.upce.compress(barcode) (invalid)', (barcode, expectedError) => {
  expect(() => upce.compress(barcode)).toThrow(expectedError)
})

test.each([
  ['127890', '012000007897'],
  ['1278907', '012000007897'],
  ['01278907', '012000007897'],
  ['01278906', '012000007896'],
  ['01238930', '012300000890'],
  ['01291940', '012910000090'],
  ['01291150', '012911000050']
])('gtin.upce.expand(%p) (expandable: %p)', (originalBarcode, expandedBarcode) => {
  expect(upce.expand(originalBarcode)).toBe(expandedBarcode)
})

test.each([
  [123, 'Barcode must be a string'],
  ['123412341', 'Barcode is not of a valid format'],
  ['123', 'Barcode is not of a valid format'],
  ['abc', 'Barcode is not of a valid format']
])('gtin.upce.expand(%p) (error: %p)', (barcode, expectedError) => {
  expect(() => upce.expand(barcode)).toThrow(expectedError)
})


lib/upce.js
'use strict'

const common = require('./common')

exports.expand = function (barcode) {
  if (typeof barcode !== 'string') throw new Error(common.NONSTRING_ERR)
  if (!/^\d{6,8}$/.test(barcode)) throw new Error(common.FORMAT_ERR)
  return setCheckDigit(
    setNumberSystem(
      expandBarcode(
        barcode
      )
    )
  )
}

function expandBarcode (barcode) {
  var digits = barcode.split('')
  var numberSystem = (digits.length > 7) ? digits.shift() : ''
  var checkDigit = (digits.length > 6) ? digits.pop() : ''
  var lastDigit = +digits.pop()
  var expanded

  switch (lastDigit) {
    case 0:
    case 1:
    case 2:
      expanded = digits.slice(0, 2).join('') + lastDigit + '0000' + digits.slice(2, 5).join('')
      break
    case 3:
    case 4:
      expanded = digits.slice(0, lastDigit).join('') + '00000' + digits.slice(lastDigit, 5).join('')
      break
    case 5:
    case 6:
    case 7:
    case 8:
    case 9:
      expanded = digits.join('') + '0000' + lastDigit
      break
  }

  return numberSystem + expanded + checkDigit
}

exports.compress = function (barcode) {
  if (typeof barcode !== 'string') throw new Error(common.NONSTRING_ERR)
  if (!/^\d{10,12}$/.test(barcode)) throw new Error(common.FORMAT_ERR)
  return compressBarcode(
    setCheckDigit(
      setNumberSystem(
        barcode
      )
    )
  )
}

function compressBarcode (barcode) {
  var v1 = /^(\d{3})([0-2])0{4}(\d{3})(\d{1})$/
  var v2 = /^(\d{3})([3-9])0{5}(\d{2})(\d{1})$/
  var v3 = /^(\d{5})0{5}(\d{1})(\d{1})$/
  var v4 = /^(\d{6})0{4}(\d{2})$/
  var rgx

  rgx = v1.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[3] + rgx[2] + rgx[4]
  }

  rgx = v2.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2] + rgx[3] + '3' + rgx[4]
  }

  rgx = v3.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2] + '4' + rgx[3]
  }

  rgx = v4.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2]
  }

  return null
}

function setCheckDigit (barcode) {
  return (
    (barcode.length !== 12 && barcode.length !== 8)
      ? barcode + common.generateCheckDigit(barcode)
      : barcode
  )
}

function setNumberSystem (barcode) {
  return (
    (barcode.length !== 12 && barcode.length !== 8)
      ? '0' + barcode
      : barcode
  )
}


lib/index.js
const common = require('./common')

function isGTIN (barcode) {
  if (typeof barcode !== 'string') throw new Error(common.NONSTRING_ERR)
  if (barcode === '') throw new Error(common.EMPTY_CODE_ERR)
  return (
    /^(\d{12,14}|\d{8})$/.test(barcode)
  )
}

function assertIsGTIN (barcode) {
  if (!isGTIN(barcode)) throw new Error(common.FORMAT_ERR)
}

function isValid (barcode) {
  assertIsGTIN(barcode)

  const checkDigit = barcode.slice(-1)
  const generatedCheckDigit = common.generateCheckDigit(barcode.slice(0, -1))

  return checkDigit === generatedCheckDigit
}

function getFormat (barcode) {
  assertIsGTIN(barcode)

  return `GTIN-${barcode.length}`
}

function minify (barcode) {
  assertIsGTIN(barcode)

  const format = +getFormat(barcode).replace('GTIN-', '')
  switch (format) {
    case 14:
      return barcode.replace(/^(0{6}|0{1,2})/, '')
    case 13:
      return barcode.replace(/^(0{5}|0{1})/, '')
    case 12:
      return barcode.replace(/^0{4}/, '')
    case 8:
      return barcode
  }
}

function getRealFormat (barcode) {
  return getFormat(minify(barcode))
}

exports.isGTIN = isGTIN
exports.isValid = isValid
exports.minify = minify
exports.getFormat = getFormat
exports.getRealFormat = getRealFormat

exports.upcE = require('./upc-e')


lib/upc-e.js
const common = require('./common')

function expand (barcode) {
  if (typeof barcode !== 'string') throw new Error(common.NONSTRING_ERR)
  if (!/^\d{6,8}$/.test(barcode)) throw new Error(common.FORMAT_ERR)
  return setCheckDigit(
    setNumberSystem(
      expandBarcode(
        barcode
      )
    )
  )
}

function expandBarcode (barcode) {
  const digits = barcode.split('')
  const numberSystem = (digits.length > 7) ? digits.shift() : ''
  const checkDigit = (digits.length > 6) ? digits.pop() : ''
  const lastDigit = +digits.pop()
  let expanded

  switch (lastDigit) {
    case 0:
    case 1:
    case 2:
      expanded = digits.slice(0, 2).join('') + lastDigit + '0000' + digits.slice(2, 5).join('')
      break
    case 3:
    case 4:
      expanded = digits.slice(0, lastDigit).join('') + '00000' + digits.slice(lastDigit, 5).join('')
      break
    case 5:
    case 6:
    case 7:
    case 8:
    case 9:
      expanded = digits.join('') + '0000' + lastDigit
      break
  }

  return numberSystem + expanded + checkDigit
}

function compress (barcode) {
  if (typeof barcode !== 'string') throw new Error(common.NONSTRING_ERR)
  if (!/^\d{10,12}$/.test(barcode)) throw new Error(common.FORMAT_ERR)
  return compressBarcode(
    setCheckDigit(
      setNumberSystem(
        barcode
      )
    )
  )
}

function compressBarcode (barcode) {
  const v1 = /^(\d{3})([0-2])0{4}(\d{3})(\d{1})$/
  const v2 = /^(\d{3})([3-9])0{5}(\d{2})(\d{1})$/
  const v3 = /^(\d{5})0{5}(\d{1})(\d{1})$/
  const v4 = /^(\d{6})0{4}(\d{2})$/
  let rgx

  rgx = v1.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[3] + rgx[2] + rgx[4]
  }

  rgx = v2.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2] + rgx[3] + '3' + rgx[4]
  }

  rgx = v3.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2] + '4' + rgx[3]
  }

  rgx = v4.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2]
  }

  return null
}

function setCheckDigit (barcode) {
  return (
    (barcode.length !== 12 && barcode.length !== 8)
      ? barcode + common.generateCheckDigit(barcode)
      : barcode
  )
}

function setNumberSystem (barcode) {
  return (
    (barcode.length !== 12 && barcode.length !== 8)
      ? '0' + barcode
      : barcode
  )
}

exports.expand = expand
exports.compress = compress


src/upc-e.ts
import {
  generateCheckDigit,
  NONSTRING_ERR,
  FORMAT_ERR
} from './common'

export function expand (barcode: string) {
  if (typeof barcode !== 'string') throw new Error(NONSTRING_ERR)
  if (!/^\d{6,8}$/.test(barcode)) throw new Error(FORMAT_ERR)
  return setCheckDigit(
    setNumberSystem(
      expandBarcode(
        barcode
      )
    )
  )
}

function expandBarcode (barcode: string) {
  const digits = barcode.split('')
  const numberSystem = (digits.length > 7) ? digits.shift()! : ''
  const checkDigit = (digits.length > 6) ? digits.pop()! : ''
  const lastDigit = +digits.pop()!

  let expanded

  switch (lastDigit) {
    case 0:
    case 1:
    case 2:
      expanded = digits.slice(0, 2).join('') + lastDigit + '0000' + digits.slice(2, 5).join('')
      break
    case 3:
    case 4:
      expanded = digits.slice(0, lastDigit).join('') + '00000' + digits.slice(lastDigit, 5).join('')
      break
    case 5:
    case 6:
    case 7:
    case 8:
    case 9:
      expanded = digits.join('') + '0000' + lastDigit
      break
  }

  return numberSystem + expanded + checkDigit
}

export function compress (barcode: string) {
  if (typeof barcode !== 'string') throw new Error(NONSTRING_ERR)
  if (!/^\d{10,12}$/.test(barcode)) throw new Error(FORMAT_ERR)
  return compressBarcode(
    setCheckDigit(
      setNumberSystem(
        barcode
      )
    )
  )
}

function compressBarcode (barcode: string) {
  const v1 = /^(\d{3})([0-2])0{4}(\d{3})(\d{1})$/
  const v2 = /^(\d{3})([3-9])0{5}(\d{2})(\d{1})$/
  const v3 = /^(\d{5})0{5}(\d{1})(\d{1})$/
  const v4 = /^(\d{6})0{4}(\d{2})$/
  let rgx

  rgx = v1.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[3] + rgx[2] + rgx[4]
  }

  rgx = v2.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2] + rgx[3] + '3' + rgx[4]
  }

  rgx = v3.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2] + '4' + rgx[3]
  }

  rgx = v4.exec(barcode)
  if (rgx) {
    return rgx[1] + rgx[2]
  }

  return null
}

function setCheckDigit (barcode: string) {
  return (
    (barcode.length !== 12 && barcode.length !== 8)
      ? barcode + generateCheckDigit(barcode)
      : barcode
  )
}

function setNumberSystem (barcode: string) {
  return (
    (barcode.length !== 12 && barcode.length !== 8)
      ? '0' + barcode
      : barcode
  )
}


lib/common.js
const NONSTRING_ERR = 'Barcode must be a string'
const EMPTY_CODE_ERR = 'Barcode must not be an empty string'
const FORMAT_ERR = 'Barcode is not of a valid format'

function to13Digits (barcode) {
  return barcode.padStart(13, '0')
}

// This assumes the barcode does not have a check digit,
// AND has a number system on it.
function generateCheckDigit (barcode) {
  return String(
    (
      10 - (
        (
          to13Digits(barcode)
            .split('')
            .map((num, idx) => (
              (+num) * ((idx % 2 === 0) ? 3 : 1)
            ))
            .reduce((prev, cur) => prev + cur)
        ) % 10
      )
    ) % 10
  )
}

exports.NONSTRING_ERR = NONSTRING_ERR
exports.EMPTY_CODE_ERR = EMPTY_CODE_ERR
exports.FORMAT_ERR = FORMAT_ERR

exports.generateCheckDigit = generateCheckDigit


jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/tmp/jest_rs",

  // Automatically clear mock calls and instances between every test
  // clearMocks: false,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  coverageDirectory: 'coverage',

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  // coverageProvider: "babel",

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "json",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest',

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: 'node'

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
}


src/index.ts
import {
  generateCheckDigit,

  NONSTRING_ERR,
  EMPTY_CODE_ERR,
  FORMAT_ERR
} from './common'
import * as upcE from './upc-e'

export { upcE }

export function isGTIN (barcode: string) {
  if (typeof barcode !== 'string') throw new Error(NONSTRING_ERR)
  if (barcode === '') throw new Error(EMPTY_CODE_ERR)
  return (
    /^(\d{12,14}|\d{9})$/.test(barcode)
  )
}

function assertIsGTIN (barcode: string): asserts barcode is string {
  if (!isGTIN(barcode)) throw new Error(FORMAT_ERR)
}

export function isValid (barcode: string) {
  assertIsGTIN(barcode)

  const checkDigit = barcode.slice(-1)
  const generatedCheckDigit = generateCheckDigit(barcode.slice(0, -1))

  return checkDigit === generatedCheckDigit
}

export function getFormat (barcode: string) {
  assertIsGTIN(barcode)

  return `GTIN-${barcode.length}`
}

export function minify (barcode: string): string {
  assertIsGTIN(barcode)

  const format = getFormat(barcode)
  switch (format) {
    case 'GTIN-14':
      return barcode.replace(/^(0{6}|0{1,2})/, '')
    case 'GTIN-13':
      return barcode.replace(/^(0{5}|0{1})/, '')
    case 'GTIN-12':
      return barcode.replace(/^0{4}/, '')
    case 'GTIN-8':
      return barcode
  }

  throw new Error('Unknown barcode format')
}

export function getRealFormat (barcode: string) {
  return getFormat(minify(barcode))
}
